statistics/cmsketch.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"math"
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/spaolacci/murmur3"
+)
+
+// CMSketch is used to estimate point queries.
+// Refer: https://en.wikipedia.org/wiki/Count-min_sketch
+type CMSketch struct {
+	depth int32
+	width int32
+	count uint64
+	table [][]uint32
+}
+
+func newCMSketch(d, w int32) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	return &CMSketch{depth: d, width: w, table: tbl}
+}
+
+func (c *CMSketch) insert(val types.Datum) error {
+	bytes, err := codec.EncodeValue(nil, val)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	c.count++
+	h1, h2 := murmur3.Sum128(bytes)
+	for i := range c.table {
+		j := (h1 + h2*uint64(i)) % uint64(c.width)
+		c.table[i][j]++
+	}
+	return nil
+}
+
+type uint32arr []uint32
+
+func (a uint32arr) Len() int           { return len(a) }
+func (a uint32arr) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a uint32arr) Less(i, j int) bool { return a[i] < a[j] }
+
+func (c *CMSketch) query(val types.Datum) (uint32, error) {
+	bytes, err := codec.EncodeValue(nil, val)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	h1, h2 := murmur3.Sum128(bytes)
+	vals := make([]uint32, c.depth)
+	min := uint32(math.MaxUint32)
+	for i := range c.table {
+		j := (h1 + h2*uint64(i)) % uint64(c.width)
```

Comment:
how about extracting `(h1 + h2*uint64(i)) % uint64(c.width)` as a function ?

Reply:
Is it too simple to be a function?